src/absil/illib.fs

```diff
@@ -432,6 +432,45 @@ module List =
     let existsSquared f xss = xss |> List.exists (fun xs -> xs |> List.exists (fun x -> f x))
     let mapiFoldSquared f z xss =  mapFoldSquared f z (xss |> mapiSquared (fun i j x -> (i,j,x)))
 
+module ResizeArray =
+
+    /// Split a ResizeArray into an array of smaller chunks.
+    /// This requires `items/chunkSize` Array copies of length `chunkSize` if `items/chunkSize % 0 = 0`,
+    /// otherwise `items/chunkSize + 1` Array copies.
+    let chunkBySize chunkSize (items: ResizeArray<'t>) =
+        // we could use Seq.chunkBySize here, but that would involve many enumerator.MoveNext() calls that we can sidestep with a bit of math
+        let itemCount = items.Count
+        if itemCount = 0
+        then [||]
+        else
+            let chunksCount =
+                match itemCount / chunkSize with
+                | n when itemCount % chunkSize = 0 -> n
+                | n -> n + 1 // any remainder means we need an additional chunk to store it
+
+            [| for index in 0..chunksCount-1 do
+                let startIndex = index * chunkSize
+                let takeCount = min (itemCount - startIndex) chunkSize
+
+                // pre-creating the array and doing a single copy is slightly better than
+                // two copies via ResizeArray<'t>.GetRange(start, count).ToArray().
+                let holder = Array.zeroCreate takeCount
+                items.CopyTo(startIndex, holder, 0, takeCount)
+                yield holder |]
+
+    /// Split a large ResizeArray into a series of array chunks that are each under the
+    /// Large Object Heap limit, in order to prevent the entire array from not being garbage-collected.
+    let mapToSmallArrayChunks f (inp: ResizeArray<'t>) =
+        let itemSizeBytes = sizeof<'t>
+        let lohSizeThresholdBytes = 85_000
+        // rounding down here is good because it ensures we don't go over
+        let maxArrayItemCount = lohSizeThresholdBytes / itemSizeBytes
+
+        /// chunk the provided input into arrays that are smaller than the LOH limit
+        /// in order to prevent long-term storage of those values
+        chunkBySize maxArrayItemCount inp
+        |> Array.map (Array.map f)
```

Comment:
Please apply `f` during the chunking process itself, so there is no allocation of intermediate arrays

Reply:
I'm curious what you mean here. I should change `yield holder` => `yield holder |> Array.map f` instead? That will still incur a total of `2 * NUM_CHUNKS` array allocations for the copies, right? One for the `holder`, and another for the `Array.map f`. Unless you mean the problem point is the  `Array.map |> Array.map f` usage in `mapToSmallArrayChunks` also allocates an additional `'b [] []` to hold the chunks?

Reply:
Something like this: https://gist.github.com/dsyme/cbeb525e7c7d16ee0ad2965bf747fb4a

Reply:
ah I see what you're going for there. That's doable, I initially had something like that but was trying to go for a bulk-copy approach.  I'll make those changes forthwith.

Reply:
Oh I remember the _real_ reason I didn't go that way: we incur a bounds check on each direct access through the indexer with this pattern.  Is that a large enough worry to stay with the bulk approach?

```csharp
        // Sets or Gets the element at the given index.
        public T this[int index]
        {
            get
            {
                // Following trick can reduce the range check by one
                if ((uint)index >= (uint)_size)
                {
                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
                }
                return _items[index];
            }

            set
            {
                if ((uint)index >= (uint)_size)
                {
                    ThrowHelper.ThrowArgumentOutOfRange_IndexException();
                }
                _items[index] = value;
                _version++;
            }
        }
```

Reply:
Yes, that's better than avoiding the double-allocation and mapping.  Shame we have to pay that cost though. Maybe this is better, I dunno
```
            let mutable enum = items.GetEnumerator()
            [| for index in 0..chunksCount-1 do
                let startIndex = index * chunkSize
                let takeCount = min (itemCount - startIndex) chunkSize

                // pre-creating the array and doing a single copy is slightly better than
                // two copies via ResizeArray<'t>.GetRange(start, count).ToArray().
                let holder = Array.zeroCreate takeCount
                for i in 0 .. takeCount - 1 do
                    enum.MoveNext() |> ignore
                    holder.[i] <- f enum.Current
                yield holder |]
```

Reply:
I'm looking at the differences between the versions [on SharpLab](https://sharplab.io/#v2:DYLgZgzgNALiCGEC2AfAsAKAPQ4AQGUAHYASxl3lwCUBTCEgLxoEEAnV+AT1xIDsYA9hV4V2XXALC5k8YMBqtcAYwAWAV14BrCADpMOLLgAqKkhFysaARzUlL5gAZkaSCFlUbN+RjQe42HNxKAoQkdBJS8rwA5jAquA4eWt5MfiRSTjAubklePrgApLgADLgAvCUOUPp4AnEKAO5mNAnOru7qyfkA1LgAjH4B4sGhdHoY8uS5AEKcKS2587gAFGAgPPy4ALQAfBswAJQrbRDrtPRMQ5wAPHwwO0dlmLgvuHgNCwJqwAAmuGoQFr4aw6GZzfIqBQ0KC4ABGanIcXg5AaX1+GwAbgJgBiWkh4LxuDReGokApkQJWDoALICXEAORoAA8YMsjkpZMBzEiUQsCdISD86FlCLgmnEKHCyBFcPi4s9XpMeFkkABhL6bConHTqjQwBUvdLKly6zUlA24eoiADaKBQAF0LTQuTQLa9cErchBTeQnhh3QHXnLVMa1RryIZFvlxSo3YHcChcCIGpCRG0fYVlJ08kxyiVtnteHHA4mRLsk7hen03oYCdxLPi+ELFGSCeYPkmaDQ/vz4D8fmQSAJeLIs55LUIIIJLMrMMXcLbcGBKRshUyNiUdKDs97w1tqz8BPP3Uqp/BWDAAJK8Nd5pvM3AAKjHXSYx8VNER8E0NAzFSQfDHCqGZbNIMDnleN7Muy2bzHO/rxu6eCEJYWxKJYyJ8NElqQqIgTCH8h5YZK9AxPIyghNwZjSKQ0QqDAwDcLCn5ZIoSJFghiEvHgMCohRozmBiJCUOcPhXNcADk9w6AA4p+VAEtENDLGeF4wsEeoHDoRgCFcbLjFxH7kCo2LNnmVw6EwrACKqGFZJa36/uG74vMuigkBuqmQbeW5gRB163r04E/hmh4uQGJm/AoOjWiQ9q4NcoFSNqsWOpxXGcGE6KRWZDrhZgnrZgA6mQKjUvAopRrmqzrHcBb7EcywnGcdBiWINx3A85QWu8nzfH8AJAiCYJLJClgwvCiIqMiYpon8fBYjieJ1rgxKkuS040nSNCMiybLKJy3LTbyB0iPQQpTjQooxpKsLSpIsrIrG6VKum4Z3iqug+haRpvXqebFBaVoLnaaXus6gLzoVni7v9fqGS8wbxH9myRrB0aleFpZiqmoYZkUVUtBUpTlhxCMJhWpOVv0NbCPWLjwPeLY0G2YotLwXY9iIfYDjAQ4jsAL6aBOYGUi0ZDwYhi5uauD6AcUW5ej6+64GF6WIae4EXgFD5alB67PoT4VKsFTlw7KgFNcB72gV5OtMjBnhwRg4XIah6Es3zMQ4S0574QShECMRlCkdE5EjFR5gQLR9GMXCLEKDhBKu4YvFCCMYSCcJ1CtZc7WSdJckwApMTKV56nhlpOl6VpxufrgOWJxUFlWTZdktKb33q/G2rqoQnA6SpWvecyMKN6wMJ2/rMKd5X4WZc6fzjwmewWfiopSA6QA===)

Reply:
Ok, I'll go ahead and make that change then. Thanks again!